File: year_2023/day_07/day07.py
from __future__ import annotations

from collections import Counter
from dataclasses import dataclass

from tools import loader, parsers


@dataclass
class Hand:
    cards: list[int]
    bid: int
    hand_type: int = 0

    def __post_init__(self) -> None:
        counts = Counter(self.cards).most_common()
        if 1 in self.cards:
            jokers = self.cards.count(1)
            if jokers != 5:
                counts.remove((1, jokers))
                counts[0] = (counts[0][0], counts[0][1] + jokers)

        match counts[0][1]:
            case 5:
                self.hand_type = 6
            case 4:
                self.hand_type = 5
            case 3:
                self.hand_type = 4 if counts[1][1] == 2 else 3
            case 2:
                self.hand_type = 2 if counts[1][1] == 2 else 1

    def __lt__(self, other: Hand) -> bool:
        if self.hand_type != other.hand_type:
            return self.hand_type < other.hand_type
        for i, val in enumerate(self.cards):
            if val != other.cards[i]:
                return val < other.cards[i]
        raise ValueError('Hands are identical')


def camel_cards(data: list[str], part2: bool) -> int:
    """
    >>> print(camel_cards(parsers.lines('test.txt'), part2=False))
    6440
    >>> print(camel_cards(parsers.lines('test.txt'), part2=True))
    5905"""
    card_values = {v: k for k, v in enumerate('23456789TJQKA', start=2)}
    if part2:
        card_values['J'] = 1
    hands = [Hand(cards=[card_values[i] for i in k], bid=int(v))
             for k, v in (x.split() for x in data)]
    return sum(i * j.bid for i, j in enumerate(sorted(hands), start=1))


print(camel_cards(parsers.lines(loader.get()), part2=False))  # 249483956
print(camel_cards(parsers.lines(loader.get()), part2=True))  # 252137472

File: year_2023/day_07/day07.go
package main

import (
	"aoc/tools"
	"cmp"
	"fmt"
	"slices"
	"strconv"
	"strings"
)

type hand struct {
	cards    []int
	bid      int
	handType int
}

func (h *hand) getHandType() {
	counts := map[int]int{}

	for _, card := range h.cards {
		counts[card]++
	}

	jokers := counts[1]
	if jokers < 5 {
		delete(counts, 1)
	}

	keys := make([]int, 0, len(counts))

	for key := range counts {
		keys = append(keys, key)
	}

	slices.SortFunc(keys, func(i, j int) int {
		return cmp.Compare(counts[j], counts[i])
	})

	if jokers > 0 && jokers < 5 {
		counts[keys[0]] += jokers
	}

	switch counts[keys[0]] {
	case 5:
		h.handType = 6 // five of a kind
	case 4:
		h.handType = 5 // four of a kind
	case 3:
		if counts[keys[1]] == 2 {
			h.handType = 4 // full house
		} else {
			h.handType = 3 // three of a kind
		}
	case 2:
		if counts[keys[1]] == 2 {
			h.handType = 2 // two pairs
		} else {
			h.handType = 1 // one pair
		}
	}
}

func solve(input string, part2 bool) int {
	data := tools.ReadLines(input)
	result := 0
	cardValues := map[string]int{}
	hands := []hand{}

	for i, char := range "23456789TJQKA" {
		cardValues[string(char)] = i + 2
	}

	if part2 {
		cardValues["J"] = 1
	}

	for _, line := range data {
		values := strings.Fields(line)
		bid, _ := strconv.Atoi(values[1])
		cards := []int{}

		for _, c := range values[0] {
			cards = append(cards, cardValues[string(c)])
		}

		h := hand{cards, bid, 0}
		h.getHandType()
		hands = append(hands, h)
	}

	slices.SortFunc(hands, func(i, j hand) int {
		if i.handType != j.handType {
			return cmp.Compare(i.handType, j.handType)
		}

		for l, leftCard := range i.cards {
			rightCard := j.cards[l]
			if leftCard != rightCard {
				return cmp.Compare(leftCard, rightCard)
			}
		}

		return 0
	})

	for i, hand := range hands {
		result += (i + 1) * hand.bid
	}

	return result
}

func main() {
	data := tools.GetData(2023, 7)

	fmt.Println(solve("test.txt", false), solve("test.txt", true))
	fmt.Println(solve(data, false), solve(data, true))
}
